NumberControl: Add TypeScript prop types

mirka opened the pull request:
Part of #35744

## What?

Adds "provisional" prop types to NumberControl.

## Why?

In the past, we've had major complications with properly typing NumberControl due to the string/number ambiguity of the values both upstream and downstream. This is why NumberControl is untyped and `@ts-nocheck`ed at the moment.

However, this has started to bottleneck TypeScript migrations for other components that inherit NumberControl props. I think it would be useful to improve the situation a bit by actually defining prop types, even if there are some contradictions that need to be `@ts-ignore`d.

I hope this reads as a constructive baby step forward, rather than a messy pile of bandaids. The benefit is that we can clearly demarcate where the type contradictions are, instead of keeping a huge untyped blackhole in our component chain.

## How?

See inline code comments for minor runtime changes.

## Testing Instructions

`npm run storybook:dev` and see the Docs view for NumberControl and UnitControl.

mirka committed:
NumberControl: Improve types

mirka committed:
Add description to InputControl classname

mirka committed:
RangeControl: Add ts-expect-errors

mirka committed:
ColorPicker: Add ts-expect-errors

mirka committed:
UnitControl: Get types from NumberControl

mirka committed:
UnitControl: Remove redundant aria-label

mirka committed:
NumberControl: Simplify controls in story

mirka commented on packages/components/src/number-control/index.tsx:
```diff
@@ -35,32 +38,36 @@ export function NumberControl(
 	};
 
-	const autoComplete = typeProp === 'number' ? 'off' : null;
+	const autoComplete = typeProp === 'number' ? 'off' : undefined;
```
⚠️ Runtime change: `null` is an invalid value for the `autoComplete` attribute.

mirka commented on packages/components/src/number-control/index.tsx:
```diff
@@ -105,10 +118,13 @@ export function NumberControl(
 		 * Handles drag to update events
 		 */
 		if ( type === inputControlActionTypes.DRAG && isDragEnabled ) {
+			// @ts-expect-error TODO: Investigate
 			const [ x, y ] = payload.delta;
+			// @ts-expect-error TODO: Investigate
 			const enableShift = payload.shiftKey && isShiftStepEnabled;
```
These two lines do not error in my IDE, but error when I do `tsc --build` by CLI. Just me? 🤔

ciampo replied:
Same for me. Maybe it's a mismatch between the TS version?

Anyway, I think that the error is "correct" — [`event` is of type `SyntheticEvent`](https://github.com/WordPress/gutenberg/blob/fba021a263c2ad8c063e999e779773cb754ccb14/packages/components/src/input-control/reducer/actions.ts#L24) which I don't think has a `shiftKey` property?

We should first understand if the `shiftKey` prop gets set or if it's expected to be part of the `event` natively (since it's also used when the action is `DRAG`)

ciampo replied:
Actually, it looks like in #38753 I had solved the situation by [casting `action` to the `DragAction` type](https://github.com/WordPress/gutenberg/pull/38753/files#diff-1795d70ae6f25ef5ee79fa75d215a18e87de7d7bdc577b294376838325d983a8R114), which seems like a good solution.

The only thing that will still need to be fixed is that, even in that case, `payload` doesn't seem to have a `delta` prop, which means that:
- either our [`DragProps` types](https://github.com/WordPress/gutenberg/blob/8f6edb18656f150fec6a77f4005397573796a264/packages/components/src/input-control/types.ts) aren't exhaustive
- or somewhat the `delta` prop is just not there at runtime too (we should check 😅 )

mirka replied:
> Same for me. Maybe it's a mismatch between the TS version?

It wasn't for me. I even cleared the TS cache. Today, only the second line errors in my IDE. Very weird and... unsettling! 🙁

So, in my IDE, `if ( type === inputControlActionTypes.DRAG )` can successfully narrow down the action type to a `DragAction`, which semi-correctly changes the `payload` type. So I think it is theoretically possible to type the Actions correctly, and rely on type narrowing, so we don't have to typecast all over the place in the consumers.

I'd prefer to bump this to a separate task, so we can look into it better. It does look possible to type the [events](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/input-control/reducer/actions.ts) correctly with generics. How does that sound?

ciampo replied:
Agreed, let's look into this as a separate task — that way we can tackle all drag-related types

mirka replied:
Moved to #43945

mirka commented on packages/components/src/unit-control/index.tsx:
```diff
@@ -259,7 +259,6 @@ function UnforwardedUnitControl(
 	return (
 		<Root className="components-unit-control-wrapper" style={ style }>
 			<ValueInput
-				aria-label={ label }
```
⚠️ Runtime change: This `label` variable is a `ReactNode`, but `aria-label` only accepts strings. I went ahead and removed this redundant `aria-label`, because the same value is added on line 270 via `label` prop.

ciampo replied:
Noted. Basically the `input` will be labelled via an actual `label` element (with the `for` attribute), instead of the `aria-label` attribute 👌

mirka commented on packages/components/src/unit-control/types.ts:
```diff
@@ -71,14 +66,8 @@ export type UnitSelectControlProps = Pick< InputControlProps, 'size' > & {
 	units?: WPUnitControlUnit[];
 };
 
-// TODO: when available, should (partially) extend `NumberControl` props.
```
Yay 🧹

ciampo commented:
Going to leave a review at later / early next week, but in the meantime I'm going to leave a link to my previous attempt https://github.com/WordPress/gutenberg/pull/38753 as it may be useful

ciampo commented on packages/components/src/number-control/types.ts:
```diff
@@ -0,0 +1,80 @@
+
+export type NumberControlProps = Omit<
+	InputControlProps,
+	'isDragEnabled' | 'min' | 'max' | 'required' | 'type' | 'value'
```
Is there a reason why we can't re-use theses props from `InputControlProps`, and we re-define them instead?

mirka replied:
Good question! Some of them, because we need a different type (`min`, `max`, `value`). Almost all of them, because we need a different `@default` value. I haven't yet found a way to "import" the JSDoc description only, or override the `@default` only 😕

ciampo replied:
Yeah, that's unfortunate.

Would it make sense to still re-use the type for `InputControlProps` for those props? E.g.

`isDragEnabled?: InputControlProps[ 'isDragEnabled' ];`

mirka replied:
Yeah, I think that might be better 👍

ciampo commented on packages/components/src/unit-control/types.ts:
```diff
@@ -71,14 +66,8 @@ export type UnitSelectControlProps = Pick< InputControlProps, 'size' > & {
-	> & {
-		__unstableStateReducer?: StateReducer;
-		__unstableInputWidth?: CSSProperties[ 'width' ];
+	NumberControlProps & {
```
Should we only `Pick` the properties that we're interested in?

mirka replied:
Fixed in f759d3585e453b333b179d5dafce0d7182ceaad8 and 6000e4becd55a72ebf238fe8c93c7bdb896f7bb0

And good news, some custom CSS in UnitControl can be removed by using the `hideHTMLArrows` prop on NumberControl 🎉 I'll do this in a separate PR.

ciampo commented on packages/components/src/number-control/types.ts:
```diff
@@ -0,0 +1,80 @@
+	 *
+	 * @default 1
+	 */
+	step?: InputHTMLAttributes< HTMLInputElement >[ 'step' ] | 'any';
```
Since `step` is a property of the `input` HTML element, we should either:

- not define it at all (and just inherit it from the `input` element base props)
- if we want to override the native prop definition, we should probably do it in `InputControl`

WDYT ?

mirka replied:
Mmm, this is a hard one. I'm leaning toward defining it (or *documenting it*, more like) explicitly, and in NumberControl. Mainly because the `step` prop is very relevant to common usage patterns of NumberControl specifically, so it's helpful to surface it in Docs with a good explanation.

Another minor reason is because NumberControl explicitly sets a default step value of `1` regardless of the input `type`, which is good to document because apparently the `time` type has a [built-in default of `60`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/step#syntax). Though, I don't really care about this one because NumberControl arrow key stepping doesn't really work properly with `type`s other than `number` anyway. (Why is `type` even exposed? 🤔)

Maybe another way we can go is to remove the definition and add stories with usage examples. But even with additional stories, I think I prefer to have it in the Docs props table anyway.

ciampo replied:
Cool, makes sense

ciampo commented on packages/components/src/number-control/index.tsx:
```diff
@@ -35,32 +38,36 @@ export function NumberControl(
+	ref: ForwardedRef< any >
 ) {
 	const isStepAny = step === 'any';
+	// @ts-expect-error step should be a number but could be string
```
In my previous attempt, I ended up writing a few utils (`numberToString`, `stringToNumber`, `ensureString`, `ensureNumber`) that could be useful. For this specific line, [I had swapped `parseFloat` with `ensureNumber`](https://github.com/WordPress/gutenberg/pull/38753/files#diff-1795d70ae6f25ef5ee79fa75d215a18e87de7d7bdc577b294376838325d983a8R49)

mirka replied:
Perfect, thanks! I replaced all the `parseFloat`s, but I'll pass on the `ensureString` for this PR since that hinges on the "what type should `value` be" decision, and it's dangerous enough to affect unit test results 😄

ciampo commented on packages/components/src/number-control/index.tsx:
```diff
@@ -35,32 +38,36 @@ export function NumberControl(
 	const baseStep = isStepAny ? 1 : parseFloat( step );
 	const baseValue = roundClamp( 0, min, max, baseStep );
-	const constrainValue = ( value, stepOverride ) => {
+	const constrainValue = ( value: number, stepOverride?: number | null ) => {
```
Minor cleanup: since `stepOverride` is already an optional prop, we could avoid the `null` type and just make it `stepOverride?: number` ?

mirka replied:
Good call 👍

ciampo commented on packages/components/src/number-control/index.tsx:
```diff
@@ -35,32 +38,36 @@ export function NumberControl(
+	 * @return The updated state to apply to InputControl
 	 */
-	const numberControlStateReducer = ( state, action ) => {
+	const numberControlStateReducer = (
+		/** State from InputControl. */
+		state: InputState,
+		/** Action triggering state change. */
+		action: inputControlActionTypes.InputAction
+	) => {
```
We could also type `numberControlStateReducer` as `NumberControlProps[ '__unstableStateReducer' ]`, which would remove the need to type `state` and `action` individually

mirka replied:
Makes sense 👍

ciampo commented on packages/components/src/number-control/index.tsx:
```diff
@@ -74,10 +81,12 @@ export function NumberControl(
 			type === inputControlActionTypes.PRESS_UP ||
 			type === inputControlActionTypes.PRESS_DOWN
 		) {
+			// @ts-expect-error TODO: Investigate if this is wrong
```
This can be fixed by casting `event` to be a keyboard event (afterall, the action type is "press up" or "press down")

```tsx
const enableShift =
	( event as KeyboardEvent | undefined )?.shiftKey &&
	isShiftStepEnabled;
```

mirka replied:
Done 👍

ciampo commented on packages/components/src/number-control/index.tsx:
```diff
@@ -88,14 +97,18 @@ export function NumberControl(
 			}
 
 			if ( type === inputControlActionTypes.PRESS_UP ) {
+				// @ts-expect-error TODO: Investigate if this is wrong
 				nextValue = add( nextValue, incrementalValue );
 			}
 
 			if ( type === inputControlActionTypes.PRESS_DOWN ) {
+				// @ts-expect-error TODO: Investigate if this is wrong
 				nextValue = subtract( nextValue, incrementalValue );
 			}
 
+			// @ts-expect-error TODO: Investigate if this is wrong
 			nextState.value = constrainValue(
+				// @ts-expect-error TODO: Investigate if this is wrong
```
All of these will mostly depend on the decision that we make regarding the type of the `value` prop

ciampo reviewed (commented):
Good job! I left a few comments, a good chunk of which are inspired by the previous work in #38753.

Regarding the migration to TypeScript, the main part that still need to be addressed is which type we should assign to the `value` prop. I wrote a good recap of the situation in https://github.com/WordPress/gutenberg/pull/40535#issuecomment-1112049865. Although I would understand if we preferred to defer that decision to a separate PR, for the sake of getting this PR merged quickly and unblock further work that depends on `NumberControl` being typed

ciampo commented on packages/components/src/number-control/stories/index.js:
```diff
@@ -12,13 +12,10 @@ export default {
 		onChange: { action: 'onChange' },
+		prefix: { control: { type: 'text' } },
+		suffix: { control: { type: 'text' } },
+		type: { control: { type: 'text' } },
```
Should we prevent `value`'s prop from having a control?

mirka committed:
Simplify `stepOverride` type

mirka committed:
Add string number utils

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

mirka committed:
Replace `parseFloat` with `ensureNumber`

mirka committed:
Simplify `numberControlStateReducer` type

mirka committed:
Cast keyboard event

mirka committed:
Disable `value` control in story

mirka committed:
UnitControl: Remove `children` from types

Fixup of https://github.com/WordPress/gutenberg/pull/43474/files#diff-b21e5b36bcbb96cd162206795bc262d1ecb0b2ccb1bac9f0096d13c4a1c9c03a

mirka committed:
Make the `step` prop visible in Docs view

mirka committed:
UnitControl: Remove irrelevant props

mirka committed:
UnitControl: Fixup controls

mirka commented:
> Regarding the migration to TypeScript, the main part that still need to be addressed is which type we should assign to the `value` prop. I wrote a good recap of the situation in [#40535 (comment)](https://github.com/WordPress/gutenberg/pull/40535#issuecomment-1112049865). Although I would understand if we preferred to defer that decision to a separate PR, for the sake of getting this PR merged quickly and unblock further work that depends on `NumberControl` being typed

Yes exactly, the main motivation of this PR is to unblock our way to clearing out the `exclude`s in tsconfig, and to get Storybook docs to an acceptable state. I don't intend to solve the `value` puzzle at this time, sorry 😅

ciampo commented on packages/components/src/unit-control/types.ts:
```diff
@@ -71,42 +66,21 @@ export type UnitSelectControlProps = Pick< InputControlProps, 'size' > & {
-		/**
-		 * The children elements.
-		 */
-		children?: ReactNode;
```
This looks like it was never supposed to be there? Interesting that it was never spotted before 😅  The benefits of a TypeScript refactor!

mirka replied:
It was a side effect of a Lodash refactor, just two weeks ago 😆
https://github.com/WordPress/gutenberg/pull/43474/files#diff-6bc8b54c48ea7869398475f3f56904b911610cd33bcdec2be1563707f9657879

ciampo commented on packages/components/src/range-control/index.tsx:
```diff
@@ -138,7 +138,8 @@ function UnforwardedRangeControl< IconProps = unknown >(
 
-	const handleOnChange = ( next: string ) => {
+	const handleOnChange = ( next?: string ) => {
+		// @ts-expect-error TODO: Investigate if this is problematic
```
Should this use `ensureNumber` ?

mirka replied:
This one can't be fixed by `ensureNumber` because the problem is that `next` can potentially be undefined.

Technically, we can remove the ts-ignore without changing effective runtime behavior by doing something like this:

```jsx
let nextValue = typeof next === 'undefined' ? NaN : parseFloat( next );
```

But then again, I really think this one needs further investigation because it's unclear whether it's intended for `setValue()` to sometimes receive a NaN. It could actually be problematic, so I wouldn't want to imply that this behavior is legit.

ciampo commented on packages/components/src/number-control/index.tsx:
```diff
@@ -0,0 +1,209 @@
+				const applyEmptyValue =
+					required === false && currentValue === '';
+
+				// @ts-expect-error TODO: Investigate if this is wrong
```
Instead of "Investigate if this is wrong", should the comment be more about the fact that this error will only be solved once we make a decision on the `value` type?

(same for the rest of the `@ts-expect-error`  comments about the type of the `value` prop)

mirka replied:
Good call, I improved the TODO comments all around. 90948f0

One of the interesting causes was the [`isValueEmpty()`](https://github.com/WordPress/gutenberg/blob/90948f0b1c9c1cace590e39d436336c74ffb57ed/packages/components/src/utils/values.js#L24) util function. I was able to fix the typing, but it's a bit out of scope for this PR so I'll submit it separately.

ciampo commented on packages/components/src/number-control/index.tsx:
```diff
@@ -0,0 +1,209 @@
+			 * Handles drag to update events
+			 */
+			if ( type === inputControlActionTypes.DRAG && isDragEnabled ) {
+				// @ts-expect-error TODO: Investigate
+				const [ x, y ] = payload.delta;
+				// @ts-expect-error TODO: Investigate
+				const enableShift = payload.shiftKey && isShiftStepEnabled;
```
Let's leave more descriptive comments about the `expect-error` reason

mirka replied:
Done in 90948f0

mirka committed:
Use InputControlProps where possible

mirka committed:
Improve TODO comments

ciampo reviewed (approved):
🚀 

Good to be merged once we add a CHANGELOG entry

mirka committed:
Add changelog

mirka committed:
Fix e2e test
